clippy_lints/src/functions.rs

```diff
@@ -144,35 +205,96 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Functions {
         self.check_line_number(cx, span, body);
     }
 
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx hir::Item) {
+        let attr_span = must_use_span(&item.attrs);
+        if let hir::ItemKind::Fn(ref decl, ref _header, ref _generics, ref body_id) = item.node {
+            self.depth += 1;
+            if let Some(span) = attr_span {
+                check_must_use_unit(cx, decl, item.span, span);
+                return;
+            }
+            if self.depth == 1 && !returns_unit(decl) && !in_external_macro(cx.sess(), item.span) {
+                check_pure_must_use(
+                    cx,
+                    cx.tcx.hir().body(*body_id),
+                    item.span,
+                    "this looks like a pure function without `#[must_use]` attribute",
+                );
+            }
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx hir::ImplItem) {
+        if let hir::ImplItemKind::Method(ref sig, ref body_id) = item.node {
+            let attr_span = must_use_span(&item.attrs);
+            self.depth += 1;
+            if let Some(span) = attr_span {
+                check_must_use_unit(cx, &sig.decl, item.span, span);
+            } else if self.depth == 1 && !returns_unit(&sig.decl) && !in_external_macro(cx.sess(), item.span) {
+                check_pure_must_use(
+                    cx,
+                    cx.tcx.hir().body(*body_id),
+                    item.span,
+                    "this looks like a pure method without `#[must_use]` attribute",
+                );
+            }
+        }
+    }
+
     fn check_trait_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx hir::TraitItem) {
         if let hir::TraitItemKind::Method(ref sig, ref eid) = item.node {
             // don't lint extern functions decls, it's not their fault
             if sig.header.abi == Abi::Rust {
                 self.check_arg_number(cx, &sig.decl, item.span);
             }
 
+            let attr_span = must_use_span(&item.attrs);
+            if let Some(span) = attr_span {
+                check_must_use_unit(cx, &sig.decl, item.span, span);
+            }
+            self.depth += 1;
             if let hir::TraitMethod::Provided(eid) = *eid {
                 let body = cx.tcx.hir().body(eid);
                 self.check_raw_ptr(cx, sig.header.unsafety, &sig.decl, body, item.hir_id);
+
+                if attr_span.is_none()
+                    && self.depth == 1
+                    && !returns_unit(&sig.decl)
+                    && !in_external_macro(cx.sess(), item.span)
+                {
+                    check_pure_must_use(
+                        cx,
+                        body,
+                        item.span,
+                        "this looks like a pure method without `#[must_use]` attribute",
```

Comment:
The message here is repeated and could be in a `const`.

Reply:
I changed the code so the message is no longer duplicated.